src/modules/sdbus/watcher.c

```diff
@@ -0,0 +1,156 @@
+/************************************************************\
+ * Copyright 2023 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+/* watcher.c - a flux watcher that becomes ready when sd-bus needs service
+ */
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <poll.h>
+#include <errno.h>
+#include <time.h>
+#include <flux/core.h>
+#include <math.h>
+#include <systemd/sd-bus.h>
+
+#include "watcher.h"
+
+struct sdbus_watcher {
+    sd_bus *bus;
+    flux_watcher_t *in;
+    flux_watcher_t *out;
+    flux_watcher_t *tmout;
+    flux_watcher_t *prep;
+    flux_watcher_t *check;
+
+    flux_watcher_t *w;
+
+    flux_watcher_f cb;
+    void *cb_arg;
+};
+
+/* called just before poll(2) call */
+static void prep_cb (flux_reactor_t *r,
+                     flux_watcher_t *w,
+                     int revents,
+                     void *arg)
+{
+    struct sdbus_watcher *sdw = arg;
+    int events;
+    uint64_t usec;
+    struct timespec ts;
+
+    flux_watcher_stop (sdw->in);
+    flux_watcher_stop (sdw->out);
+    flux_watcher_stop (sdw->tmout);
+
+    if ((events = sd_bus_get_events (sdw->bus)) >= 0) {
+        if ((events & POLLIN))
+            flux_watcher_start (sdw->in);
+        if ((events & POLLOUT))
+            flux_watcher_start (sdw->out);
+    }
+    if (sd_bus_get_timeout (sdw->bus, &usec) >= 0
+        && usec != UINT64_MAX
+        && clock_gettime (CLOCK_MONOTONIC, &ts) == 0) {
+        double now = floor (1E6*ts.tv_sec + 1E-3*ts.tv_nsec); // to usec
+        double timeout = 1E-6*(usec - now);
+
+        if (timeout >= 0.) {
+            flux_timer_watcher_reset (sdw->tmout, timeout, 0.);
+            flux_watcher_start (sdw->tmout);
+        }
+    }
+}
+
+/* called just after poll(2) return */
+static void check_cb (flux_reactor_t *r,
+                      flux_watcher_t *w,
+                      int revents,
+                      void *arg)
+{
+    struct sdbus_watcher *sdw = arg;
+
+    if (sdw->cb)
+        sdw->cb (r, sdw->w, FLUX_POLLIN, sdw->cb_arg);
```

Comment:
this initially threw me for a loop, until i realized the `revents` here doesn't really matter.  Could use a comment.  Also, perhaps could use POLLOUT instead of POLLIN, given this is the subscription callback.

Reply:
I'm redoing this that the fd/timer watcher revents is just passed through to the callback.  I'm hoping that won't require any comment.  It seems clear that revents is not useful in the callback - you have to call `sd_bus_process()` regardless.